Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,348 @@
+        '\t        print("If called")'
+    )
+}
+
+
+def create_tmp_python_file(source_code, dir=None):
+    """Common utility function for writing source code to temporary file.
+
+    It creates file with `.py` extension.
+
+    Returns: Absolute path of created code file."""
+    with NamedTemporaryFile(mode='w', dir=dir,
+                            delete=False, suffix=".py") as f:
+        f.write(source_code)
+    return f.name
+
+
+class TestFormatWitnesses(TestCase):
+    """Testing `tabnanny.format_witnesses()`."""
+
+    def test_format_whitenesses(self):
+        """Asserting formatter result by giving various input samples."""
+        tests = [
+            {'words': 'Test' , 'expected': 'at tab sizes T, e, s, t'},
+            {'words': ''     , 'expected': 'at tab size '},
+            {'words': 't'    , 'expected': 'at tab size t'},
+            {'words': '  t  ', 'expected': 'at tab sizes  ,  , t,  ,  '},
+        ]
+
+        for test in tests:
+            self.assertEqual(
+                tabnanny.format_witnesses(test['words']), test['expected']
+            )
+
+class TestErrPrint(TestCase):
+    """Testing `tabnanny.errprint()`."""
+
+    def test_errprint(self):
+        """Asserting result of `tabnanny.errprint()` by giving sample inputs."""
+        tests = [
+            {'args': ['first', 'second'], 'expected': 'first second\n'},
+            {'args': ['first']          , 'expected': 'first\n'},
+            {'args': [1, 2, 3]          , 'expected': '1 2 3\n'},
+            {'args': []                 , 'expected': '\n'}
+        ]
+
+        for test in tests:
+            with captured_stderr() as stderr:
+                tabnanny.errprint(*test['args'])
+            self.assertEqual(stderr.getvalue() , test['expected'])
+
+
+class TestNannyNag(TestCase):
+    """Testing `tabnanny.NannyNag` exception."""
+    def setUp(self):
+        self.tests = [
+            {
+                'nanny': tabnanny.NannyNag(0, "foo", "bar"),
+                'expected': {
+                    'lineno': 0,
+                    'msg': 'foo',
+                    'line': 'bar'
+                }
+            },
+            {
+                'nanny': tabnanny.NannyNag(5, "testmsg", "testline"),
+                'expected': {
+                    'lineno': 5,
+                    'msg': 'testmsg',
+                    'line': 'testline'
+                }
+            }
+        ]
+
+    def test_all_methods(self):
+        """Asserting behaviour of `tabnanny.NannyNag` exception."""
+        for test in self.tests:
+            nanny = test['nanny']
+            self.assertEqual(test['expected']['lineno'], nanny.get_lineno())
+            self.assertEqual(test['expected']['msg'], nanny.get_msg())
+            self.assertEqual(test['expected']['line'], nanny.get_line())
+
+
+class TestCheck(TestCase):
+    """Testing tabnanny.check()."""
+
+    def setUp(self):
+        tabnanny.verbose = 0  # Forcefully deactivating verbose mode.
```

Comment:
Unit tests must have zero side effect: you must save the old verbose value and restore it. Example:

```
self.addCleanup(setattr, tabnanny, 'verbose', tabnanny.verbose)
tabnanny.verbose = 0
```